src/include/rho/AllocationTable.hpp

```diff
@@ -0,0 +1,326 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, a copy is available at
+ *  https://www.R-project.org/Licenses/
+ */
+
+#ifndef RHO_ALLOCATIONTABLE_HPP
+#define RHO_ALLOCATIONTABLE_HPP
+
+#include <assert.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <cstdint>
+#include <cstdlib>
+#include <functional>
+#include <limits>
+#include <map>
+
+#include "rho/AllocatorSuperblock.hpp"
+
+namespace rho {
+  /*
+   * Special-purpose hashtable implementation for tracking medium and large
+   * allocations.
+   *
+   * The table uses open addressing with quadratic probing to resolve
+   * collisions. This hashtable differs from dense_hash_table by
+   * allowing asymmetric insertion and searching - allocation ranges are
+   * inserted but the lookup function searches on a pointer so the search
+   * will look for a range covering the pointer.
+   *
+   * Another difference from most other types of generic maps is that this
+   * hashtable will map a single allocation to multiple hash buckets - the
+   * reason being that each hash bucket represents addresses that hash into
+   * that bucket, and all internal pointers of an allocation must be able
+   * to be found via the lookup_pointer() member function.
+   *
+   * For hashing pointers, we shift away LOW_BITS bits from the pointer before
+   * hashing, and this limits the range of hash keys one allocation can hash
+   * to.
+   */
+  class AllocationTable {
+  private:
+    class Allocation;
+
+  public:
+
+    /**
+     * \brief Allocates a new allocation table of the give size.
+     *
+     * @param num_bits the number of bits to use for hashing. The number of
+     * buckets is 2^num_bits.
+     */
+    AllocationTable(unsigned num_bits);
+
+    /**
+     * \brief Frees the memory used by the hashtable, but not the allocations in it.
+     */
+    ~AllocationTable();
+
+    /**
+     * Iterates over all current allocations and calls the argument function on
+     * their pointers.
+     *
+     * It is okay to free objects during the iteration, but adding objects
+     * means they may not be found during the current iteration pass.
+     */
+    void applyToAllAllocations(std::function<void(void*)> fun);
+
+    /**
+     * Erases all hashtable entries for a pointer. The size determines
+     * how many hash buckets are removed.
+     */
+    void erase(uintptr_t pointer, size_t size);
+
+    /**
+     * Remove an allocation from the hashtable.  If the allocation is in a
+     * superblock, the superblock is left as is but the superblock bitset is
+     * updated to indicate that the block is free.
+     * If the allocation is a large object (ie. not in a superblock), then
+     * the object is taken out of the hashtable and inserted into a freelist.
```

Comment:
All of this stuff about superblocks etc doesn't really belong in the AllocationTable, but rather in the code that uses it.  

Presumably you'll need to add something to the api to let the caller know if an allocation is a superblock or not.